opal/mca/base/mca_base_source.c

```diff
@@ -0,0 +1,242 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2013-2018 Los Alamos National Security, LLC. All rights
+ *                         reserved.
+ * Copyright (c) 2015      Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2015      Bull SAS.  All rights reserved.
+ * Copyright (c) 2015      The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * Copyright (c) 2017      IBM Corporation. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "mca_base_source.h"
+#include "mca_base_vari.h"
+
+#include <stddef.h>
+#include <sys/time.h>
+#include <sys/resource.h>
+
+#include "opal/class/opal_pointer_array.h"
+#include "opal/class/opal_hash_table.h"
+#include "opal/util/clock_gettime.h"
+
+static opal_pointer_array_t registered_sources;
+static bool mca_base_source_initialized = false;
+static int source_count;
+
+static opal_mutex_t mca_base_source_lock = OPAL_MUTEX_STATIC_INIT;
+
+int mca_base_source_default_source = -1;
+
+static uint64_t mca_base_source_default_time_source (void)
+{
+    uint64_t time_value;
+    struct timespec current;
+
+    (void)opal_clock_gettime(&current);
+    time_value = 1000000000ul * current.tv_sec + current.tv_nsec;
+
+    return time_value;
+}
+
+static uint64_t mca_base_source_default_time_source_ticks (void)
+{
+    struct timespec spec;
+    if (0 == opal_clock_getres(&spec)){
+        return (uint64_t)(spec.tv_sec + spec.tv_nsec);
+    } else {
+        /* guess */
+        return 1000000;
+    }
+}
+
+/***************************************************************************************************/
+
+int mca_base_source_init (void)
+{
+    int ret = OPAL_SUCCESS;
+
+    OPAL_THREAD_LOCK(&mca_base_source_lock);
+    if (false == mca_base_source_initialized) {
+        mca_base_source_initialized = true;
+
+        OBJ_CONSTRUCT(&registered_sources, opal_pointer_array_t);
+        opal_pointer_array_init(&registered_sources, 16, 512, 16);
```

Comment:
Should we make these constants be MCA params (perhaps level 9)? I don't have _too_ strong of an opinion here, but it is odd to see hard-coded constants in the middle of C code.

Reply:
This appears to be an issue with several of these arrays, such as in opal/mca/base/mca_base_pvar.c :
```
opal_pointer_array_init(&registered_pvars, 128, 2048, 128);
```
The current source values would definitely address current use.
Would it make sense to address this more generally in a different PR?

Reply:
Yeah, I guess if this is something systemic and outside the scope of just MPI_T, then it's probably worth a future PR.